VDA-DID-Linkage/contracts/VeridaDIDLinkage.sol

```diff
@@ -0,0 +1,256 @@
+
+        string memory strDID = StringsUpgradeable.toHexString(uint256(uint160(didAddr)));
+        strDID = string(abi.encodePacked("did:vda:", strDID));
+
+        require(bytes(_identifierController[info.identifier]).length == 0, "Controller existing");
+
+        // Verify data
+        {
+            bytes memory params = abi.encodePacked(
+                didAddr,
+                info.identifier,
+                info.signedData,
+                info.signedProof
+            );
+
+            verifyRequest(didAddr, params, requestSignature, requestProof);
+
+            params = abi.encodePacked(
+                strDID,
+                "|",
+                info.identifier
+            );
+
+            if (StringUtils.equal(_identifierType[kind], "Self")) {
+                // SignatureProof should be signed by the id of identifier
+                address[] memory signers = new address[](1);
+                signers[0] = parseAddr(id);
+
+                verifyDataWithSigners(params, info.signedData, info.signedProof, signers);
+
+            } else {
+                // SignatureProof should be signed by the trusted signers
+                verifyData(params, info.signedData, info.signedProof);
+            }
+        }
+
+        // Link did & identifier
+        _identifierController[info.identifier] = strDID;
+        _didIdentifiers[strDID].add(info.identifier);
+
+        emit Link(strDID, info.identifier);
+    }
+
+    /**
+     * @dev See {IVeridaDIDLinkage}
+     */
+    function unlink(
+        address didAddr, 
+        string calldata identifier,
+        bytes calldata requestSignature,
+        bytes calldata requestProof
+    ) external override {
+        string memory strDID = StringsUpgradeable.toHexString(uint256(uint160(didAddr)));
+        strDID = string(abi.encodePacked("did:vda:", strDID));
+
+        require(_didIdentifiers[strDID].contains(identifier), "Unlinked identifier");
+
+        // Verify request
+        {
+            bytes memory params = abi.encodePacked(
+                strDID,
+                "|",
+                identifier
+            );
+
+            verifyRequest(didAddr, params, requestSignature, requestProof);
+        }
+
+        delete _identifierController[identifier];
+        _didIdentifiers[strDID].remove(identifier);
+
+        emit Unlink(strDID, identifier);
+    }
+
+    /**
+     * @dev See {IVeridaDIDLinkage}
+     */
+    function isLinked(string calldata did, string calldata identifier) external view override returns(bool) {
+        return _didIdentifiers[did].contains(identifier);
+    }
+
+    /**
+     * @dev See {IVeridaDIDLinkage}
+     */
+    function getController(string calldata identifier) external view override returns(string memory) {
+        return _identifierController[identifier];
+    }
+
+    /**
+     * @dev See {IVeridaDIDLinkage}
+     */
+    function getIdentifierList(string calldata did) external view override returns(string[] memory) {
```

Comment:
Rename to `links(did)`

Reply:
Updated

Reply:
How about to rename to getLinks(did)?
Because, there is a function named `link()` in the contract.
There could be a bit confusion for the users if we defined 2 functions of `link()` & `links()`.